src/app/components/shopping-list/shopping-list.component.ts

```diff
@@ -1,55 +1,92 @@
 
+	async ionViewDidEnter(): Promise<void> {
+		await this.loadPantryList()
+	}
+
 	async loadPantryList(): Promise<void> {
 		this.pantryList$ = await this.storage.get_all(true, false)
 		const adHoc = await this.storage.get_all(false, true)
 		adHoc.subscribe(adHocEntryList => {
-			this.inCart = adHocEntryList
+			this.adHocItems = adHocEntryList.filter(([, item]: AdHocShoppingEntry) => !item.inCart)
+			this.inCart = adHocEntryList.filter(([, item]: AdHocShoppingEntry) => item.inCart)
 		})
 
 		this.pantryList$.subscribe(pantryList => {
-			this.runningLow = pantryList.filter(([, pantryItem]) => pantryItem.runningLow && !pantryItem.inCart)
 			this.outOf = pantryList.filter(([, pantryItem]) => pantryItem.out && !pantryItem.inCart)
+			this.runningLow = pantryList.filter(([, pantryItem]) => pantryItem.runningLow && !pantryItem.inCart)
 			this.inCart = this.inCart.concat(pantryList.filter(([, pantryItem]) => pantryItem.inCart))
 		})
+
+		this.inCart.forEach(([key]) => {
+			this.inCartCheckedStatusByKey[key] = true
+		})
 	}
 
-	async ngOnInit(): Promise<void> {
-		await this.loadPantryList()
+	placeInList(entry: PantryEntry | AdHocShoppingEntry, listName: string): void {
+		const allListNames = ['outOf', 'runningLow', 'adHocItems', 'inCart']
+		const allListNamesButTarget = allListNames.filter(name => name != listName)
+		this[listName].push(entry)
+		for (const name of allListNamesButTarget) {
+			this[name] = this[name].filter(([key]) => key != entry[0])
+		}
+	}
+	async _toggleCartStatus([key, item]: AdHocShoppingEntry | PantryEntry, newState: boolean): Promise<void> {
+		item.inCart = newState
+		this.inCartCheckedStatusByKey[key] = newState
+		await this.storage.update(key, item)
+	}
+	async placeInCart([key, item]: PantryEntry | AdHocShoppingEntry): Promise<void> {
+		this.placeInList([key, item], 'inCart')
+		await this._toggleCartStatus([key, item], true)
 	}
 
-	async ionViewDidEnter(): Promise<void> {
-		await this.loadPantryList()
+	async uncheckCartItem([key, item]: PantryEntry | AdHocShoppingEntry): Promise<void> {
+		let targetListName: string
+		if (this.storage.isAdHoc(key)) {
+			targetListName = 'adHocItems'
+		} else {
+			if ((item as PantryItem).runningLow) {
+				targetListName = 'runningLow'
+			} else {
+				targetListName = 'outOf'
+			}
+		}
+		this.placeInList([key, item], targetListName)
+		this._toggleCartStatus([key, item], false)
 	}
 
-	placeInCart(entry: PantryEntry): void {
-		this.runningLow = this.runningLow.filter(([key]) => entry[0] !== key)
-		this.outOf = this.outOf.filter(([key]) => entry[0] !== key)
-		this.inCart.push(entry)
-		entry[1].inCart = true
-		this.storage.update(...entry)
+	async emptyCheckedItemsFromCart(): Promise<void> {
+		for (const [key, item] of this.inCart.filter(([key]) => this.inCartCheckedStatusByKey[key])) {
+			if (this.storage.isAdHoc(key)) {
+				this.storage.delete(key)
+			} else {
+				;(item as PantryItem).runningLow = false
```

Comment:
This (`;(`) is the price we pay for not using semicolons at the end of lines like civilized people?

Reply:
its actually not necessary haha, but prettier insists on it.  eslint says to remove them!  false cannot be a valid function name, so false() has no chance of needing a semicolon, but i think prettier just insists on protecting any line starting with ()